crates/dojo-lang/src/model.rs

```diff
@@ -1,15 +1,139 @@
+use crate::plugin::{DojoAuxData, Model, DOJO_MODEL_ATTR};
+
+const CURRENT_MODEL_VERSION: u8 = 1;
+const MODEL_VERSION_NAME: &str = "version";
+
+/// Get the version associated with the dojo::model attribute.
+///
+/// Note: dojo::model attribute has already been checked so there is one and only one attribute.
+///
+/// Parameters:
+/// * db: The semantic database.
+/// * struct_ast: The AST of the model struct.
+/// * diagnostics: vector of compiler diagnostics.
+///
+/// Returns:
+/// * The model version associated with the dojo:model attribute.
+pub fn get_model_version(
+    db: &dyn SyntaxGroup,
+    struct_ast: ItemStruct,
+    diagnostics: &mut Vec<PluginDiagnostic>,
+) -> u8 {
+    if let OptionArgListParenthesized::ArgListParenthesized(arguments) =
+        struct_ast.attributes(db).query_attr(db, DOJO_MODEL_ATTR).first().unwrap().arguments(db)
+    {
+        let version_args = arguments
+            .arguments(db)
+            .elements(db)
+            .iter()
+            .filter_map(|a| match a.arg_clause(db) {
+                ArgClause::Named(x) => {
+                    let arg_name = x.name(db).text(db).to_string();
+                    if arg_name.eq(MODEL_VERSION_NAME) {
+                        Some(x.value(db))
+                    } else {
+                        diagnostics.push(PluginDiagnostic {
+                            message: format!("Unexpected argument '{}' for dojo::model", arg_name),
+                            stable_ptr: x.stable_ptr().untyped(),
+                            severity: Severity::Warning,
+                        });
+                        None
+                    }
+                }
+                ArgClause::Unnamed(x) => {
+                    diagnostics.push(PluginDiagnostic {
+                        message: format!(
+                            "Unexpected argument '{}' for dojo::model",
+                            x.as_syntax_node().get_text(db)
+                        ),
+                        stable_ptr: x.stable_ptr().untyped(),
+                        severity: Severity::Warning,
+                    });
+                    None
+                }
+                ArgClause::FieldInitShorthand(x) => {
+                    diagnostics.push(PluginDiagnostic {
+                        message: format!(
+                            "Unexpected argument '{}' for dojo::model",
+                            x.name(db).name(db).text(db).to_string()
+                        ),
+                        stable_ptr: x.stable_ptr().untyped(),
+                        severity: Severity::Warning,
+                    });
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let version = match version_args.len().cmp(&1) {
+            Ordering::Equal => match version_args.first().unwrap() {
+                Expr::Literal(v) => {
+                    if let Ok(int_value) = v.text(db).parse::<u8>() {
```

Comment:
Would it be better to parse a `u64` here, and then the validation function of each parameter can make the range check of its choice?

Reply:
`dojo::model` arguments have been filtered in a previous code block, so here we process the `version` argument only. We can extend the version type from `u8` to `u64` but I don't know if we need it at the moment 🤔 

We could replace this `get_model_version` by a more general function like `parse_arguments` which would return a `HashMap` of <ArgumentName,ArgumentValue>. This function would handle duplicated arguments and would handle `String` value, `u64` value and "flag argument" like `#[my_attr(fast)]`. Then, with the returned `HashMap` we can check ranges, type of value, etc ... for each argument.

But:
- at the moment, we have only one argument which is `version` and I would keep this simple and indicates to the user unexpected arguments,
- There is a plan to migrate dojo-lang to use [Scarb proc-macro](https://github.com/software-mansion/scarb/blob/main/design/01-proc-macro.md), so some code refactoring will be probably done. 

What do you think ?

Reply:
> dojo::model arguments have been filtered in a previous code block, so here we process the version argument only. We can extend the version type from u8 to u64 but I don't know if we need it at the moment

Oh my bad, it's good then let's keep that for this iteration that's great. Let's also keep `u8` as you mention.

> at the moment, we have only one argument which is version and I would keep this simple and indicates to the user unexpected arguments,

We will soon need a new attribute `layout` to handle different storage layouts. So once this one is merged, we could adapt to add an other one.

> There is a plan to migrate dojo-lang to use [Scarb proc-macro](https://github.com/software-mansion/scarb/blob/main/design/01-proc-macro.md), so some code refactoring will be probably done.

100% agree.